lib/IDE/IDETypeChecking.cpp

```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
   return Scratch;
 }
 
+/// This walker will traverse the AST and report types for every variable declaration.
+class VariableTypeCollector: public SourceEntityWalker {
+private:
+  SourceManager &SM;
+  unsigned int BufferId;
+  std::vector<VariableTypeInfo> &Results;
+
+  /// This is where we print all types.
+  llvm::raw_ostream &OS;
+
+  /// Map from a printed type to the offset in OS where the type starts.
+  llvm::StringMap<uint32_t> TypeOffsets;
+
+  /// Specified by the client whether we should canonicalize types before printing.
+  const bool CanonicalType;
+
+  /// Find an exicisting offset in the type buffer, otherwise print the type.
```

Comment:
This comment actually confused me for a bit. I think the following would be better:
```cpp
/// Returns the start and end offset of this string in `OS`. If it does not exist in `OS` yet, `PrintedType` will be added to `OS`.
```